src/Detector.php

```diff
@@ -67,19 +67,26 @@ public function determineMimeType(\SplFileObject $file): string
     /**
      * Read width and height from a image file.
      *
-     * @param string $filePath
+     * @param \SplFileObject $file File object
      * @return array [$width, $height]
      * @throws \InvalidArgumentException Unreadable SplFileInfo
      */
-    public function determineDimensions(string $filePath): array
+    public function determineDimensions(\SplFileObject $file): array
     {
-        $stream = new \SplFileInfo($filePath);
-        if (!$stream->isReadable()) {
-            throw new \InvalidArgumentException('File path ' . $filePath . ' is not existed');
+        if (false !== $path = $file->getRealPath()) {
+            $info = getimagesize($path);
+
+            return [$info[0], $info[1]];
+        }
+
+        $file->rewind();
+        $buffer = '';
+        while (false === $file->eof()) {
+            $buffer .= $file->fread(512);
```

Comment:
[`SplFileObject::fread`](https://secure.php.net/manual/en/splfileobject.fread.php) will mutate the object as it moves the cursor.

@ancarda is that OK? If not, before rewinding the stream this method needs to check on the cursor position and reset to it after reading. That’s what `Detector::read` does, specifically see lines 79 and 86–89:

https://github.com/ancarda/file-detector/blob/d6c9042dfb331ccd57d36a227f51689e97ac1932/src/Detector.php#L77-L91

Reply:
Hmm, yeah, ideally we can reverse any mutations we've made so the caller doesn't need to keep track of state. That would apply for any exceptions too; we may need to catch exceptions so we can cleanup before rethrowing.

Reply:
@Zegnat, can you present the proper way about that you mention?
And please consider following code snippets:

```php
$this->read($file, 0, 512);
```

Is this correct to replace the code on line `85` with the previous code snippets?

Thanks.

Reply:
> That would apply for any exceptions too; we may need to catch exceptions so we can cleanup before rethrowing.

Luckily neither [`SplFileObject::fread`](https://secure.php.net/manual/en/splfileobject.eof.php) nor [`SplFileObject::fread`](https://secure.php.net/manual/en/splfileobject.fread.php) throw exceptions. There should never be an unexpected exit during the copying of the stream contents to a string. And if [`SplFileObject::rewind`](https://secure.php.net/manual/en/splfileobject.rewind.php) throws that happens before any mutations (i.e. the first mutation we try fails).

> And please consider following code snippets:
>
> ```php
> $this->read($file, 0, 512);
> ```
>
> Is this correct to replace the code on line `85` with the previous code snippets?

Sadly, no, that doesn’t work. The `Detector::read` method always grabs the bytes from a specific point forward. In the case of that snippet it will start at byte 0 (start of the file) and read the first 512 bytes. You could use it by keeping track of how much has been read and updating the second argument (0) accordingly, but that seems like more work than necessary.

I also think you would incur some performance penalties calling `Detector::read` over and over again. Every time the method will do the work to reset the cursor position. That really only needs to happen once: after we’ve finished copying. (This isn’t really an issue for `Detector::determineMimeType`, which does a maximum of 2 reads.)

> can you present the proper way about that you mention?

The following code is untested, but it should give you the general idea:

```php
$position = $file->ftell();
if (0 !== $position) {
    $file->rewind();
}
$buffer = '';
while (false === $file->eof()) {
    $buffer .= $file->fread(512);
}
if (false === $position || -1 === $file->fseek($position, \SEEK_SET)) {
    throw new \Exception('Could not reset the cursor.');
}
```

Basically it does the following:

1. Store the current cursor position.
2. If the cursor isn’t at the start of the file (position 0): rewind.
3. While the cursor hasn’t reached the end of the file, read 512 bytes at a time into an in-memory string. The 512 is an arbitrary number. There might be better values. I haven’t done the research.
4. If the cursor can’t be placed back to where it was because the position is faulty or the seeking failed, throw an exception.

As I said, it is untested, so make sure to have unit tests in place if you copy any of that code. I also haven’t thought about the `Detector::read` method since I wrote it 1½ months ago, so if you see any possible enhancements (performance or otherwise), definitely contribute!

Reply:
@Zegnat, thank you for your reply. And I also follow your suggestion to implement this feature :+1: .

Please review them. If having any request changes, feel free to write the comments for that :smile: .